crypto/dilithium/p_pqdsa_test.cc

```diff
@@ -1108,232 +1108,323 @@ TEST_P(PQDSAParameterTest, KAT) {
 }
 
 TEST_P(PQDSAParameterTest, KeyGen) {
-  // Basic key generation tests for MLDSA
-  // Generate a MLDSA key
-  int nid = GetParam().nid;
-  size_t pk_len = GetParam().public_key_len;
-  size_t sk_len = GetParam().private_key_len;
-
+  // ---- 1. Test basic key generation flow ----
+  // Create context of PQDSA type and a key pair
   bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_PQDSA, nullptr));
   ASSERT_TRUE(ctx);
+
+  // Setup the context with specific PQDSA parameters.
+  int nid = GetParam().nid;
   ASSERT_TRUE(EVP_PKEY_CTX_pqdsa_set_params(ctx.get(),nid));
-  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+
+  // Generate a key pair.
   EVP_PKEY *raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
   ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
   bssl::UniquePtr<EVP_PKEY> pkey(raw);
 
-  // Extract public key and check it is of the correct size
-  uint8_t *buf = nullptr;
-  size_t buf_size;
-  EXPECT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
-  EXPECT_EQ(pk_len, buf_size);
-
-  buf = (uint8_t *)OPENSSL_malloc(buf_size);
-  ASSERT_NE(buf, nullptr);
-  EXPECT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
-
-  buf_size = 0;
-  EXPECT_FALSE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
+  // ---- 2. Test key generation with PKEY as a template ----
+  ctx.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+  ASSERT_TRUE(ctx);
 
+  // Generate a key pair.
+  raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+  ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
+  pkey.reset(raw);
+
+  // ---- 3. Test getting raw keys and their size ----
+  size_t pk_len, sk_len;
+
+  // First getting the sizes only.
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), nullptr, &pk_len));
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(pkey.get(), nullptr, &sk_len));
+  EXPECT_EQ(pk_len, GetParam().public_key_len);
+  EXPECT_EQ(sk_len, GetParam().private_key_len);
+
+  // Then getting the keys and the sizes.
+  std::vector<uint8_t> pk_raw(pk_len);
+  std::vector<uint8_t> sk_raw(sk_len);
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), pk_raw.data(), &pk_len));
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(pkey.get(), sk_raw.data(), &sk_len));
+  EXPECT_EQ(pk_len, GetParam().public_key_len);
+  EXPECT_EQ(sk_len, GetParam().private_key_len);
+
+  // ---- 4. Test failure modes for EVP_PKEY_CTX_pqdsa_set_params. ----
+  // ctx is NULL.
+  ASSERT_FALSE(EVP_PKEY_CTX_pqdsa_set_params(nullptr, nid));
   uint32_t err = ERR_get_error();
   EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
-  EXPECT_EQ(EVP_R_BUFFER_TOO_SMALL, ERR_GET_REASON(err));
-  OPENSSL_free(buf);
-  buf = nullptr;
+  EXPECT_EQ(ERR_R_PASSED_NULL_PARAMETER, ERR_GET_REASON(err));
```

Comment:
If you want, these three lines that repeat can be made into a macro. I think all of them are ERR_LIB_EVP.

Reply:
I'm not sure I understand exactly what you mean here. Open to suggestions.

Reply:
These 3 lines
```
 uint32_t err = ERR_get_error();
  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
  EXPECT_EQ(ERR_R_PASSED_NULL_PARAMETER, ERR_GET_REASON(err));
```
repeat throughout the tests and can be replaced by a single-line macro that takes the reason as a parameter (similar to the CMP macros).
It will condense the code a bit.

Reply:
Okay, see https://github.com/aws/aws-lc/pull/2062/commits/463aa0d675b49755a0f796ed8827483a52bd0bd2#diff-d043ba12b97b57a0d69c31db07b10cee8ca8c05cf41e1875981b254bc7cbe670L1186-L1189

Reply:
[16f48d6](https://github.com/aws/aws-lc/pull/2062/commits/16f48d60912aa7e259f5fd809ba0f90abaa1c976) done